File: app/src/main/java/com/najmi/oreamnos/model/GenerationPill.java
package com.najmi.oreamnos.model;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/**
 * Represents a Custom Refinement Pill - a user-defined refinement command.
 * Pills appear as selectable chips alongside built-in refinements (Rephrase,
 * etc.).
 */
public class GenerationPill {

    private String id;
    private String name; // Display label (e.g., "Punchy")
    private String command; // AI instruction (e.g., "Make it punchy and engaging")

    // Required for Gson deserialization
    public GenerationPill() {
        this.id = UUID.randomUUID().toString();
    }

    public GenerationPill(String name, String command) {
        this.id = UUID.randomUUID().toString();
        this.name = name;
        this.command = command;
    }

    // Getters
    public String getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public String getCommand() {
        return command;
    }

    // Setters
    public void setId(String id) {
        this.id = id;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setCommand(String command) {
        this.command = command;
    }

    /**
     * Converts a list of pills to JSON string.
     */
    public static String toJson(List<GenerationPill> pills) {
        return new Gson().toJson(pills);
    }

    /**
     * Parses JSON string to list of pills.
     */
    public static List<GenerationPill> fromJson(String json) {
        if (json == null || json.trim().isEmpty()) {
            return new ArrayList<>();
        }
        try {
            Type listType = new TypeToken<List<GenerationPill>>() {
            }.getType();
            List<GenerationPill> pills = new Gson().fromJson(json, listType);
            return pills != null ? pills : new ArrayList<>();
        } catch (Exception e) {
            return new ArrayList<>();
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;
        GenerationPill that = (GenerationPill) o;
        return id != null && id.equals(that.id);
    }

    @Override
    public int hashCode() {
        return id != null ? id.hashCode() : 0;
    }
}

File: app/src/main/java/com/najmi/oreamnos/model/GenerationPill.kt
package com.najmi.oreamnos.model

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.reflect.TypeToken
import java.lang.reflect.Type
import java.util.UUID

/**
 * Represents a Custom Refinement Pill - a user-defined refinement command.
 * Pills appear as selectable chips alongside built-in refinements (Rephrase, etc.).
 */
data class GenerationPill(
    var id: String = UUID.randomUUID().toString(),
    var name: String = "",
    var command: String = ""
) {
    /**
     * Secondary constructor for Java compatibility (name and command only).
     */
    constructor(name: String, command: String) : this(
        id = UUID.randomUUID().toString(),
        name = name,
        command = command
    )

    companion object {
        // Cache the Gson instance with the custom adapter to avoid recreation overhead
        private val gson: Gson by lazy {
            GsonBuilder()
                .registerTypeAdapter(GenerationPill::class.java, Deserializer())
                .create()
        }

        /**
         * Converts a list of pills to JSON string.
         */
        @JvmStatic
        fun toJson(pills: List<GenerationPill>): String {
            return gson.toJson(pills)
        }

        /**
         * Parses JSON string to list of pills.
         * Uses a custom deserializer to ensure default values are respected if fields are missing.
         */
        @JvmStatic
        fun fromJson(json: String?): List<GenerationPill> {
            if (json.isNullOrBlank()) return emptyList()
            return try {
                val listType = object : TypeToken<List<GenerationPill>>() {}.type
                gson.fromJson<List<GenerationPill>>(json, listType) ?: emptyList()
            } catch (e: Exception) {
                emptyList()
            }
        }

        /**
         * Custom Deserializer to handle missing fields in JSON.
         * Ensures that non-nullable fields like 'id' are properly initialized with defaults
         * even if the JSON object is empty or partial.
         */
        private class Deserializer : JsonDeserializer<GenerationPill> {
            override fun deserialize(
                json: JsonElement,
                typeOfT: Type,
                context: JsonDeserializationContext
            ): GenerationPill {
                val obj = json.asJsonObject

                // Safely extract fields, falling back to defaults if missing or null
                val id = if (obj.has("id") && !obj.get("id").isJsonNull) {
                    obj.get("id").asString
                } else {
                    UUID.randomUUID().toString()
                }

                val name = if (obj.has("name") && !obj.get("name").isJsonNull) {
                    obj.get("name").asString
                } else {
                    ""
                }

                val command = if (obj.has("command") && !obj.get("command").isJsonNull) {
                    obj.get("command").asString
                } else {
                    ""
                }

                return GenerationPill(id, name, command)
            }
        }
    }
}
